pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java

```diff
@@ -0,0 +1,362 @@
+import org.apache.pinot.segment.spi.index.reader.RangeIndexReader;
+import org.apache.pinot.segment.spi.index.reader.SortedIndexReader;
+import org.apache.pinot.segment.spi.index.reader.TextIndexReader;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+import org.apache.pinot.segment.spi.store.ColumnIndexType;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.spi.config.table.BloomFilterConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.spi.data.FieldSpec.DataType;
+
+
+public final class PhysicalColumnIndexContainer implements ColumnIndexContainer {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PhysicalColumnIndexContainer.class);
+
+  private final ForwardIndexReader<?> _forwardIndex;
+  private final InvertedIndexReader<?> _invertedIndex;
+  private final RangeIndexReader<?> _rangeIndex;
+  private final TextIndexReader _textIndex;
+  private final TextIndexReader _fstIndex;
+  private final JsonIndexReader _jsonIndex;
+  private final H3IndexReader _h3Index;
+  private final BaseImmutableDictionary _dictionary;
+  private final BloomFilterReader _bloomFilter;
+  private final NullValueVectorReaderImpl _nullValueVectorReader;
+
+  public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, ColumnMetadata metadata,
+      IndexLoadingConfig indexLoadingConfig, File segmentIndexDir)
+      throws IOException {
+    String columnName = metadata.getColumnName();
+    boolean loadInvertedIndex = indexLoadingConfig.getInvertedIndexColumns().contains(columnName);
+    boolean loadRangeIndex = indexLoadingConfig.getRangeIndexColumns().contains(columnName);
+    boolean loadTextIndex = indexLoadingConfig.getTextIndexColumns().contains(columnName);
+    boolean loadFSTIndex = indexLoadingConfig.getFSTIndexColumns().contains(columnName);
+    boolean loadJsonIndex = indexLoadingConfig.getJsonIndexColumns().contains(columnName);
+    boolean loadH3Index = indexLoadingConfig.getH3IndexConfigs().containsKey(columnName);
+    boolean loadOnHeapDictionary = indexLoadingConfig.getOnHeapDictionaryColumns().contains(columnName);
+    BloomFilterConfig bloomFilterConfig = indexLoadingConfig.getBloomFilterConfigs().get(columnName);
+
+    if (segmentReader.hasIndexFor(columnName, ColumnIndexType.NULLVALUE_VECTOR)) {
+      PinotDataBuffer nullValueVectorBuffer = segmentReader.getIndexFor(columnName, ColumnIndexType.NULLVALUE_VECTOR);
+      _nullValueVectorReader = new NullValueVectorReaderImpl(nullValueVectorBuffer);
+    } else {
+      _nullValueVectorReader = null;
+    }
+
+    if (loadTextIndex) {
+      Preconditions.checkState(segmentReader.hasIndexFor(columnName, ColumnIndexType.TEXT_INDEX));
+      Map<String, Map<String, String>> columnProperties = indexLoadingConfig.getColumnProperties();
+      _textIndex = new LuceneTextIndexReader(columnName, segmentIndexDir, metadata.getTotalDocs(),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource that might leak to a class field. This can cause a resource leak that slows down or crashes your system. For example, when an else condition at line 148 evaluates to true, the *PhysicalColumnIndexContainer* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *\_textIndex* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).